### PR TITLE
Don't derive Debug for `OnceWith` & `RepeatWith`

### PR DESCRIPTION
**Pull request by ghost**
Closures don't impl Debug, so the derived impl is kinda useless. The behavior of not debug-printing closures is consistent with the rest of the iterator adapters/sources.

### CODE REVIEW

File: library/core/src/iter/sources/once_with.rs [hunk 1]
```diff
@@ -66,12 +67,22 @@ pub fn once_with<A, F: FnOnce() -> A>(gen: F) -> OnceWith<F> {
...
+#[derive(Clone)]
 #[stable(feature = "iter_once_with", since = "1.43.0")]
 pub struct OnceWith<F> {
     gen: Option<F>,
 }
 
+impl<F> fmt::Debug for OnceWith<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.gen.is_some() {
+            f.write_str("OnceWith(Some)")
```
Comment (H4x5) [inline library/core/src/iter/sources/once_with.rs:79]: Is Some & None here ideal?

**Comment by H4x5** [comment]
What kind of stability attributes should the impls have?

**Comment by m-ou-se** [comment]
> What kind of stability attributes should the impls have?

You could use something like these:

`#[stable(feature = "iter_once_with_debug", since = "CURRENT_RUSTC_VERSION")]`

`#[stable(feature = "iter_repeat_with_debug", since = "CURRENT_RUSTC_VERSION")]`

**Comment by m-ou-se** [comment]
@rfcbot merge

This PR removes the `F: Debug` bound from `Debug for OnceWith<F>` and `Debug for RepeatWith<F>`, such that they still implement Debug when used with a closure or function.

This is consistent with e.g. `Map<I, F>` and `Filter<I, F>`, which only require `I: Debug` but not `F: Debug` for their own Debug impls.

**Comment by dtolnay** [comment]
I am marking yaahc's box because she has stepped down from T-libs-api by this point.

File: library/core/src/iter/sources/once_with.rs (hunk 1 above)
Comment (fmease) [reply library/core/src/iter/sources/once_with.rs:80]: You could print `OnceWith(Some(_))` which looks a bit better IMO.

**Review by dtolnay — APPROVED** [review]

**Comment by klensy** [comment]
https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy commits should be rebased (and optionally squashed), cc #106333

**Comment by klensy** [comment]
Sorry for noise, but now it's push-after-r+

**Pull request merged**